src/utils/SparseGrid.h

```diff
@@ -0,0 +1,170 @@
+    /*! \brief Returns all data within radius of query_pt.
+     *
+     * Finds all values with location within radius of \p query_pt.  May
+     * return additional values that are beyond radius.
+     *
+     * Average running time is a*(1 + 2 * radius / cell_size)**2 +
+     * b*cnt where a and b are proportionality constance and cnt is
+     * the number of returned items.  The search will return items in
+     * an area of (2*radius + cell_size)**2 on average.  The max range
+     * of an item from the query_point is radius + cell_size.
+     *
+     * \param[in] query_pt The point to search around.
+     * \param[in] radius The search radius.
+     * \return Vector of values found
+     */
+    std::vector<Val> getNearby(const Point &query_pt, coord_t radius) const;
+
+private:
+    using GridPoint = Point;
+    using GridMap = std::unordered_multimap<GridPoint, Val>;
+
+    /*! \brief Add values from the cell indicated by \p grid_pt.
+     *
+     * \param[in] grid_pt The grid coordinates of the cell.
+     * \param[out] ret The values in the cell are appended to ret.
+     */
+    void addFromCell(const GridPoint &grid_pt,
+                     std::vector<Val> &ret) const;
+
+    /*! \brief Add values from cells that might contain sought after points.
+     *
+     * Appends values from cell that might have values within \p
+     * radius of \p query_pt.  Appends all values that are within
+     * radius of query_pt.  May append values that are up to radius +
+     * cell_size from query_pt.
+     *
+     * \param[in] query_pt The point to search around.
+     * \param[in] radisu The search radius.
+     * \param[out] ret The values in the cell are appended to ret.
+     */
+    void addNearby(const Point &query_pt, coord_t radius,
+                   std::vector<Val> &ret) const;
+
+    /*! \brief Compute the grid coordinates of a point.
+     *
+     * \param[in] point The actual location.
+     * \return The grid coordinates that correspond to point.
+     */
+    GridPoint toGridPoint(const Point &point) const;
+
+    /*! \brief Map from grid locations (GridPoint) to values (Val). */
+    GridMap m_grid;
+    /*! \brief Accessor for getting locations from values. */
+    PointAccess m_point_access;
+    /*! \brief The cell (square) size. */
+    coord_t m_cell_size;
+};
+
+#define SG_TEMPLATE template<class Val, class PointAccess>
+#define SG_THIS SparseGrid<Val, PointAccess>
+
+SG_TEMPLATE
+SG_THIS::SparseGrid(coord_t cell_size)
+{
+    m_cell_size = cell_size;
+}
+
+SG_TEMPLATE
+typename SG_THIS::GridPoint SG_THIS::toGridPoint(const Point &point)  const
+{
+    // This mapping via truncation results in the cells with
+    // GridPoint.x==0 being twice as large and similarly for
+    // GridPoint.y==0.  This doesn't cause any incorrect behavior,
+    // just changes the running time slightly.  The change in running
+    // time from this is probably not worth doing a proper floor
+    // operation.
+    return point / m_cell_size;
```

Comment:
We had a lil discussion about this and indeed, since most printers have their origin at the side of the printer this will never really be a problem.